src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/hearing/UnlistHearingAction.java

```diff
@@ -0,0 +1,169 @@
+        ids.addAll(sessionsIds);
+
+        return ids.stream().toArray(UUID[]::new);
+    }
+
+    @Override
+    public void act() {
+        if (sessions.isEmpty()) {
+            throw new RuntimeException("Hearing parts assigned to Hearing haven't been listed yet");
+        }
+
+        originalHearingParts = mapHearingPartsToStrings(hearingParts);
+
+        hearingParts.stream().forEach(hp -> {
+            hp.setSession(null);
+            hp.setSessionId(null);
+            hp.setStart(null);
+            VersionInfo vi = getVersionInfo(hp);
+            hp.setVersion(vi.getVersion());
+        });
+
+        hearingPartRepository.save(hearingParts);
+    }
+
+    private VersionInfo getVersionInfo(HearingPart hp) {
+        Optional<VersionInfo> hpvi = unlistHearingRequest.getHearingPartsVersions()
+            .stream()
+            .filter(hpv -> hpv.getId().equals(hp.getId()))
+            .findFirst();
+        return hpvi.orElseThrow(() ->
+            new RuntimeException("Couldn't find version for hearing part with id " + hp.getId().toString())
+        );
+    }
+
+    @Override
+    public List<UserTransactionData> generateUserTransactionData() {
+        List<UserTransactionData> userTransactionDataList = new ArrayList<>();
+        originalHearingParts.forEach((id, hpString) ->
+            userTransactionDataList.add(new UserTransactionData("hearingPart",
+                id,
+                hpString,
+                "update",
+                "update",
+                0)
+            )
+        );
+
+        userTransactionDataList.add(getLockedEntityTransactionData("hearing", hearing.getId()));
+        sessions.stream().forEach(s ->
+            userTransactionDataList.add(getLockedEntityTransactionData("session", s.getId()))
+        );
+
+        return userTransactionDataList;
+    }
+
+    @Override
+    public List<FactMessage> generateFactMessages() {
+        List<FactMessage> msgs = new ArrayList<>();
+
+        hearingParts.forEach(hp -> {
+            String msg;
+            try {
+                msg = factsMapper.mapHearingToRuleJsonMessage(hp);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException(e);
+            }
+            msgs.add(new FactMessage(RulesService.UPSERT_HEARING_PART, msg));
+        });
+
+        return msgs;
+    }
+
+    @Override
+    public UUID getUserTransactionId() {
+        return unlistHearingRequest.getUserTransactionId();
+    }
+
+    private UserTransactionData getLockedEntityTransactionData(String entity, UUID id) {
+        return new UserTransactionData(entity, id, null, "lock", "unlock", 0);
+    }
+
+    private Map<UUID, String> mapHearingPartsToStrings(List<HearingPart> hearingParts) {
+        Map<UUID, String> originalIdStringPair = new HashMap<UUID, String>();
+        hearingParts.stream().forEach(hp -> {
+            try {
+                String hearingPartString = objectMapper.writeValueAsString(hp);
```

Comment:
This could probably be extracted too to avoid try catches in every place, I thought we've done this earlier, but can't find it right now.